extensions/amp-link-rewriter/0.1/link-rewriter.js

```diff
@@ -0,0 +1,187 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {getConfigOpts} from './config-options';
+import {getDataParamsFromAttributes} from '../../../src/dom';
+
+const WL_ANCHOR_ATTR = [
+  'href',
+  'id',
+  'rel',
+  'rev',
+];
+const PREFIX_DATA_ATTR = /^vars(.+)/;
+const REG_DOMAIN_URL = /^https?:\/\/(www\.)?([^\/:]*)(:\d+)?(\/.*)?$/;
+const PAGE_PROP_WHITELIST = {
+  'SOURCE_URL': true,
+  'DOCUMENT_REFERRER': true,
+};
+
+export class LinkRewriter {
+  /**
+   * @param {!AmpElement} ampElement
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampDoc
+   */
+  constructor(ampElement, ampDoc) {
+    /** @private {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = ampDoc;
+
+    /** @private {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
+
+    /** @private {?Event} */
+    this.event_ = null;
+
+    /** @private {?Object} */
+    this.configOpts_ = getConfigOpts(ampElement);
+
+    /** @public {string} */
+    this.rewrittenUrl = this.configOpts_.output;
+
+    /** @private {!../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacementService_ = Services.urlReplacementsForDoc(ampElement);
+  }
+
+  /**
+   * @param {!Event} event
+   */
+  handleClick(event) {
+    this.event_ = event;
+
+    const htmlElement = this.event_.srcElement;
+    const sourceTrimmedDomain = Services
+        .documentInfoForDoc(this.ampDoc_)
+        .sourceUrl.replace(/(www\.)?(.*)/, '$2');
+    const canonicalTrimmedDomain = Services
+        .documentInfoForDoc(this.ampDoc_)
+        .sourceUrl.replace(/(www\.)?(.*)/, '$2');
+
+    if (!htmlElement) {
+      return;
+    }
+
+    if (this.isInternalLink_(htmlElement, sourceTrimmedDomain)
+      || this.isInternalLink_(htmlElement, canonicalTrimmedDomain)) {
+      return;
+    }
+
+    this.setRedirectUrl_(htmlElement);
+  }
+
+  /**
+   * Check if the anchor element leads to an internal link
+   * @param {?Element} htmlElement
+   * @param {?string} trimmedDomain
+   * @return {boolean}
+   */
+  isInternalLink_(htmlElement, trimmedDomain) {
+    const href = htmlElement.getAttribute('href');
+
+    return !(href && REG_DOMAIN_URL.test(href) &&
+            RegExp.$2 !== trimmedDomain);
+  }
+
+  /**
+   * @param {!Element} htmlElement
+   * return {Promise}
+   */
+  setRedirectUrl_(htmlElement) {
+    const oldValHref = htmlElement.getAttribute('href');
+
+    return this.replacePageProp_().then(() => {
+      const {vars} = this.configOpts_;
+
+      if (vars instanceof Object) {
+        htmlElement.href = this.replaceVars_(htmlElement, vars);
+      }
+
+      this.viewer_.win.setTimeout(() => {
+        htmlElement.href = oldValHref;
+      }, 500);
+
+    });
+  }
+
+  /**
+   * @return {Promise}
+   */
+  replacePageProp_() {
+    return this.urlReplacementService_.expandUrlAsync(
```

Comment:
You may want to switch to `exapndUrlSync` because we want to resolve as soon as possible before navigation.

Reply:
switched to `exapndUrlSync`